### PR TITLE
feat(plugin): TextReplace

### PR DESCRIPTION
**Pull request by AutumnVN**
### Things i did:

- [x] replace message before send
- [x] better gui 
- [x] use DataStore
- [x] support both string and regex
- [x] editable
- [x] auto create new empty line

### How does this shit currently look:

image

### CODE REVIEW

**Comment by lewisakura** [comment]
I think this plugin would have a lot better UX if it wasn't just a massive textarea. Users probably won't understand very well, regardless of what you tell them at the top. I'd recommend splitting this into three textboxes instead:
- Find
- Replace
- Only if (for your conditional replacements)
and then having a +/x button to modify how many there are.

**Comment by abUwUser** [comment]
if they're planning to add more symbols, then maybe it is better to stay with how it is, or just add more custom interfaces
also is there any way to replace the url no matter what? like the only consideration is it to have, for example, `twitter.com`

**Comment by abUwUser** [comment]
i meant extra symbols as adding more characters that do something, and i mean "characters that do something" as `=>` or `?` as for example. i do not know the proper name for them

**Comment by abUwUser** [comment]
> do you want always replace `twitter.com` to something else? just `twitter.com => something`

i see

**Comment by ArjixWasTaken** [comment]
The aliucord counterpart does this better

**Comment by ArjixWasTaken** [comment]
Honestly, why does this have to edit the message before sending it?

It would make more sense to visually edit messages w/o actually editing them.

Like, I might want to send a Twitter link, but when for me to appear as a nitter link.

I might want to send gn, but when I view it to be good night, whilst the others only see gn

**Comment by ArjixWasTaken** [comment]
If someone w/o the plugin sends me `gn`, I want to see `good night`, but this plugin doesn't do that.

**Comment by Vendicated** [comment]
As others have mentioned, it'd be much better to have the settings ui render two textboxes per replace rule, with an :x: button and a ➕ button to add an additional row. Optimally, this should be provided by Vencord as OptionType.ARRAY, iirc someone is working on that

**Comment by Vendicated** [comment]
Please use settings.store instead of Settings.plugins.TextReplace everywhere! Also, I'm not a big fan of storing rules as an array with 3 elements, an object would be more intuitive

**Comment by AutumnVN** [comment]
idk how to use DataStore for the thing at line 103-147

**Comment by eggsy** [comment]
Hey guys!

Coming from #557, I could say I have a **suggestion** to this plugin. What if you could make a way to let users create their own lists for services? A simple dropdown with some categories such as:

- Social Media (TikTok, Instagram, Reddit etc.)
- Games (Steam, HumbleBundle etc.)

and that will basically add new records to your storage automatically. Basically some predefined lists users can install with a single click. **Or** you could just implement a button to activate the social media links replacers since tons of users manually replace the links of those services pretty often. Take this as a suggestion now, I'll take a look into this when I have time for this.

**Comment by abUwUser** [comment]
> Hey guys!
> 
> Coming from #557, I could say I have a **suggestion** to this plugin. What if you could make a way to let users create their own lists for services? A simple dropdown with some categories such as:
> 
> - Social Media (TikTok, Instagram, Reddit etc.)
> - Games (Steam, HumbleBundle etc.)
> 
> and that will basically add new records to your storage automatically. Basically some predefined lists users can install with a single click. **Or** you could just implement a button to activate the social media links replacers since tons of users manually replace the links of those services pretty often. Take this as a suggestion now, I'll take a look into this when I have time for this.

If i recall correctly, we had that discussion on Vencord's server on how to implement this. One of the suggestions was to make a channel at the server dedicated for text replacements. This can avoid an increase of issues/PRs requesting text replacements to be added as defaults, but it'll cog up the server with another channel. I don't know what does Ven think about that

**Comment by somerandomcloud** [comment]
Hey @AutumnVN, this looks actually awesome, but im hoping the buttons arent permanent?

**Comment by AutumnVN** [comment]
> but im hoping the buttons arent permanent?

the row will dissapear as you click x button, does that what you mean?

**Comment by somerandomcloud** [comment]
> > but im hoping the buttons arent permanent?
> 
> the row will dissapear as you click x button, is that what you mean?

No, I mean the current buttons are kind of ugly. Hoping they change to an svg that suits discords ux or smth

**Comment by AutumnVN** [comment]
hmm then what should i use, idk how it should look to suit discord

File: src/plugins/textReplace.tsx [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+                            </td>
+                            <td>
+                                <Button
+                                    color={Button.Colors.RED}
+                                    size={Button.Sizes.MIN}
+                                    onClick={() => {
+                                        settings.store.rules.splice(index, 1);
+                                        // This is to force update the table after deleting a rule
+                                        settings.store.owo = "owo";
+                                        settings.store.owo = "uwu";
```
Comment (ArjixWasTaken) [inline src/plugins/textReplace.tsx:133]: Use forceUpdater

https://github.com/Vendicated/Vencord/blob/main/src/utils/misc.tsx#L74-L77

```typescript

const update = useForceUpdater();
...

update(); // will re-render the component
```

Comment (ArjixWasTaken) [reply src/plugins/textReplace.tsx:133]: It's just semantically better.

Comment (AutumnVN) [reply src/plugins/textReplace.tsx:133]: it doesn't work, the deleted line still there like without it.

**Comment by C10udburst** [comment]
> The aliucord counterpart does this better

😎

**Comment by AutumnVN** [comment]
switched to DataStore

**Comment by AutumnVN** [comment]
is it ok now or is there anything need to be changed

**Comment by ArjixWasTaken** [comment]
Can you stop merging all the time?
Why are you not rebasing?

There is no reason to pollute the commit history with merges from upstream.

**Comment by AutumnVN** [comment]
sorry, i didnt know about rebase until now 💀

**Comment by Vendicated** [comment]
merge is preferred in a pull request to keep the history clean and and linear, we will squash upon merge

**Comment by ArjixWasTaken** [comment]
The history is full and linear, are you talking about the GitHub comments/convos getting messed up?

Or are you talking about the commit date changing?

**Comment by abUwUser** [comment]
any reason for the closure? just curious

**Pull request closed without merging**